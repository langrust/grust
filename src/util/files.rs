use codespan_reporting::files;
use std::ops::Range;

/// A file that is backed by an `Arc<String>`.
#[derive(Debug, Clone)]
struct File {
    /// The name of the file.
    name: String,
    /// The source code of the file.
    source: String,
    /// The starting byte indices in the source code.
    line_starts: Vec<usize>,
}

impl File {
    fn line_start(&self, line_index: usize) -> Result<usize, files::Error> {
        use std::cmp::Ordering;

        match line_index.cmp(&self.line_starts.len()) {
            Ordering::Less => Ok(*self
                .line_starts
                .get(line_index)
                .expect("failed despite previous check")),
            Ordering::Equal => Ok(self.source.len()),
            Ordering::Greater => Err(files::Error::LineTooLarge {
                given: line_index,
                max: self.line_starts.len() - 1,
            }),
        }
    }
}

/// An opaque file identifier.
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub struct FileId(i32);
impl FileId {
    /// Construct and return the default file identifier.
    pub fn default() -> Self {
        return FileId(-1);
    }
}

/// The files handler.
#[derive(Debug, Clone)]
pub struct Files {
    files: Vec<File>,
}

impl Files {
    /// Create a new files database.
    pub fn new() -> Files {
        Files { files: Vec::new() }
    }

    /// Add a file to the database, returning the handle that can be used to
    /// refer to it again.
    pub fn add(
        &mut self,
        name: impl Into<String>,
        source: impl Into<String>,
    ) -> Option<FileId> {
        let file_id = FileId(i32::try_from(self.files.len()).ok()?);
        let name = name.into();
        let source = source.into();
        let line_starts = files::line_starts(&source).collect();

        self.files.push(File {
            name,
            line_starts,
            source,
        });

        Some(file_id)
    }

    /// Get the file corresponding to the given id.
    fn get(&self, file_id: FileId) -> Result<&File, files::Error> {
        self.files
            .get(file_id.0 as usize)
            .ok_or(files::Error::FileMissing)
    }
}

impl<'files> files::Files<'files> for Files {
    type FileId = FileId;
    type Name = &'files str;
    type Source = &'files str;

    fn name(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(self.get(file_id)?.name.as_ref())
    }

    fn source(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(&self.get(file_id)?.source)
    }

    fn line_index(
        &self,
        file_id: FileId,
        byte_index: usize
    ) -> Result<usize, files::Error> {
        self.get(file_id)?
            .line_starts
            .binary_search(&byte_index)
            .or_else(|next_line| Ok(next_line - 1))
    }

    fn line_range(
        &self,
        file_id: FileId,
        line_index: usize,
    ) -> Result<Range<usize>, files::Error> {
        let file = self.get(file_id)?;
        let line_start = file.line_start(line_index)?;
        let next_line_start = file.line_start(line_index + 1)?;

        Ok(line_start..next_line_start)
    }
}