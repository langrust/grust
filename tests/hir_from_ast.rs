use codespan_reporting::files::SimpleFiles;

use grustine::{error::display, hir_from_ast};

#[test]
fn hir_from_ast_transformation_for_counter() {
    let mut files = SimpleFiles::new();

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/counter.gr").expect("unkown file"),
    );

    match hir_from_ast(counter_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_blinking() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/blinking.gr").expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/button_management.gr").expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management_condition_match() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/button_management_condition_match.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management_using_function() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/button_management_using_function.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_pid() {
    let mut files = SimpleFiles::new();

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/pid.gr").expect("unkown file"),
    );

    match hir_from_ast(pid_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}
