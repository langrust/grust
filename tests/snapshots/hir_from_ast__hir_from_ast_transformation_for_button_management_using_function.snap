---
source: tests/hir_from_ast.rs
expression: file
---
typedefs:
  - Enumeration:
      id: Button
      elements:
        - Released
        - Pressed
      location: ~
  - Enumeration:
      id: ResetState
      elements:
        - InProgress
        - Confirmed
      location: ~
functions:
  - id: reset_state_management
    inputs:
      - - button_state
        - Enumeration: Button
      - - counter
        - Integer
      - - reset_limit_1
        - Integer
    statements:
      - id: result
        element_type:
          Enumeration: ResetState
        expression:
          Match:
            expression:
              Call:
                id: button_state
                typing:
                  Enumeration: Button
                location: ~
            arms:
              - - Constant:
                    constant:
                      Enumeration:
                        - Button
                        - Released
                    location: ~
                - ~
                - Constant:
                    constant:
                      Enumeration:
                        - ResetState
                        - Confirmed
                    typing:
                      Enumeration: ResetState
                    location: ~
              - - Constant:
                    constant:
                      Enumeration:
                        - Button
                        - Pressed
                    location: ~
                - Application:
                    function_expression:
                      Call:
                        id: " >= "
                        typing:
                          Abstract:
                            - - Integer
                              - Integer
                            - Boolean
                        location: ~
                    inputs:
                      - Call:
                          id: counter
                          typing: Integer
                          location: ~
                      - Call:
                          id: reset_limit_1
                          typing: Integer
                          location: ~
                    typing: Boolean
                    location: ~
                - Constant:
                    constant:
                      Enumeration:
                        - ResetState
                        - InProgress
                    typing:
                      Enumeration: ResetState
                    location: ~
              - - Default:
                    location: ~
                - ~
                - Constant:
                    constant:
                      Enumeration:
                        - ResetState
                        - Confirmed
                    typing:
                      Enumeration: ResetState
                    location: ~
            typing:
              Enumeration: ResetState
            location: ~
        location: ~
    returned:
      - Enumeration: ResetState
      - Call:
          id: result
          typing:
            Enumeration: ResetState
          location: ~
    location: ~
nodes:
  - id: counter
    is_component: false
    inputs:
      - - res
        - Boolean
      - - inc
        - Integer
    unscheduled_equations:
      o:
        scope: Output
        id: o
        signal_type: Integer
        expression:
          FunctionApplication:
            function_expression:
              Call:
                id: if_then_else
                typing:
                  Abstract:
                    - - Boolean
                      - Integer
                      - Integer
                    - Integer
                location: ~
            inputs:
              - SignalCall:
                  signal:
                    id: res
                    scope: Input
                  typing: Boolean
                  location: ~
                  dependencies: ~
              - Constant:
                  constant:
                    Integer: 0
                  typing: Integer
                  location: ~
                  dependencies: ~
              - FunctionApplication:
                  function_expression:
                    Call:
                      id: " + "
                      typing:
                        Abstract:
                          - - Integer
                            - Integer
                          - Integer
                      location: ~
                  inputs:
                    - FunctionApplication:
                        function_expression:
                          Call:
                            id: (_)
                            typing:
                              Abstract:
                                - - Integer
                                - Integer
                            location: ~
                        inputs:
                          - FollowedBy:
                              constant:
                                Integer: 0
                              expression:
                                SignalCall:
                                  signal:
                                    id: o
                                    scope: Output
                                  typing: Integer
                                  location: ~
                                  dependencies: ~
                              typing: Integer
                              location: ~
                              dependencies: ~
                        typing: Integer
                        location: ~
                        dependencies: ~
                    - SignalCall:
                        signal:
                          id: inc
                          scope: Input
                        typing: Integer
                        location: ~
                        dependencies: ~
                  typing: Integer
                  location: ~
                  dependencies: ~
            typing: Integer
            location: ~
            dependencies: ~
        location: ~
    unitary_nodes: {}
    location: ~
    graph: ~
  - id: reset_button
    is_component: false
    inputs:
      - - button_state
        - Enumeration: Button
      - - period
        - Integer
      - - reset_limit_1
        - Integer
      - - reset_limit_2
        - Integer
    unscheduled_equations:
      counter:
        scope: Local
        id: counter
        signal_type: Integer
        expression:
          NodeApplication:
            node: counter
            inputs:
              - SignalCall:
                  signal:
                    id: res
                    scope: Local
                  typing: Boolean
                  location: ~
                  dependencies: ~
              - SignalCall:
                  signal:
                    id: period
                    scope: Input
                  typing: Integer
                  location: ~
                  dependencies: ~
            signal: o
            typing: Integer
            location: ~
            dependencies: ~
        location: ~
      res:
        scope: Local
        id: res
        signal_type: Boolean
        expression:
          FunctionApplication:
            function_expression:
              Call:
                id: " && "
                typing:
                  Abstract:
                    - - Boolean
                      - Boolean
                    - Boolean
                location: ~
            inputs:
              - FunctionApplication:
                  function_expression:
                    Call:
                      id: (_)
                      typing:
                        Abstract:
                          - - Boolean
                          - Boolean
                      location: ~
                  inputs:
                    - FunctionApplication:
                        function_expression:
                          Call:
                            id: " == "
                            typing:
                              Abstract:
                                - - Enumeration: Button
                                  - Enumeration: Button
                                - Boolean
                            location: ~
                        inputs:
                          - SignalCall:
                              signal:
                                id: button_state
                                scope: Input
                              typing:
                                Enumeration: Button
                              location: ~
                              dependencies: ~
                          - Constant:
                              constant:
                                Enumeration:
                                  - Button
                                  - Pressed
                              typing:
                                Enumeration: Button
                              location: ~
                              dependencies: ~
                        typing: Boolean
                        location: ~
                        dependencies: ~
                  typing: Boolean
                  location: ~
                  dependencies: ~
              - FunctionApplication:
                  function_expression:
                    Call:
                      id: (_)
                      typing:
                        Abstract:
                          - - Boolean
                          - Boolean
                      location: ~
                  inputs:
                    - FollowedBy:
                        constant:
                          Boolean: true
                        expression:
                          FunctionApplication:
                            function_expression:
                              Call:
                                id: " == "
                                typing:
                                  Abstract:
                                    - - Enumeration: Button
                                      - Enumeration: Button
                                    - Boolean
                                location: ~
                            inputs:
                              - SignalCall:
                                  signal:
                                    id: button_state
                                    scope: Input
                                  typing:
                                    Enumeration: Button
                                  location: ~
                                  dependencies: ~
                              - Constant:
                                  constant:
                                    Enumeration:
                                      - Button
                                      - Released
                                  typing:
                                    Enumeration: Button
                                  location: ~
                                  dependencies: ~
                            typing: Boolean
                            location: ~
                            dependencies: ~
                        typing: Boolean
                        location: ~
                        dependencies: ~
                  typing: Boolean
                  location: ~
                  dependencies: ~
            typing: Boolean
            location: ~
            dependencies: ~
        location: ~
      user_reset_request:
        scope: Output
        id: user_reset_request
        signal_type: Boolean
        expression:
          Match:
            expression:
              SignalCall:
                signal:
                  id: button_state
                  scope: Input
                typing:
                  Enumeration: Button
                location: ~
                dependencies: ~
            arms:
              - - Constant:
                    constant:
                      Enumeration:
                        - Button
                        - Released
                    location: ~
                - ~
                - []
                - FollowedBy:
                    constant:
                      Boolean: false
                    expression:
                      SignalCall:
                        signal:
                          id: user_reset_request
                          scope: Output
                        typing: Boolean
                        location: ~
                        dependencies: ~
                    typing: Boolean
                    location: ~
                    dependencies: ~
              - - Constant:
                    constant:
                      Enumeration:
                        - Button
                        - Pressed
                    location: ~
                - ~
                - []
                - FunctionApplication:
                    function_expression:
                      Call:
                        id: " >= "
                        typing:
                          Abstract:
                            - - Integer
                              - Integer
                            - Boolean
                        location: ~
                    inputs:
                      - SignalCall:
                          signal:
                            id: counter
                            scope: Local
                          typing: Integer
                          location: ~
                          dependencies: ~
                      - FunctionApplication:
                          function_expression:
                            Call:
                              id: " + "
                              typing:
                                Abstract:
                                  - - Integer
                                    - Integer
                                  - Integer
                              location: ~
                          inputs:
                            - SignalCall:
                                signal:
                                  id: reset_limit_1
                                  scope: Input
                                typing: Integer
                                location: ~
                                dependencies: ~
                            - SignalCall:
                                signal:
                                  id: reset_limit_2
                                  scope: Input
                                typing: Integer
                                location: ~
                                dependencies: ~
                          typing: Integer
                          location: ~
                          dependencies: ~
                    typing: Boolean
                    location: ~
                    dependencies: ~
            typing: Boolean
            location: ~
            dependencies: ~
        location: ~
      user_reset_request_state:
        scope: Output
        id: user_reset_request_state
        signal_type:
          Enumeration: ResetState
        expression:
          FunctionApplication:
            function_expression:
              Call:
                id: reset_state_management
                typing:
                  Abstract:
                    - - Enumeration: Button
                      - Integer
                      - Integer
                    - Enumeration: ResetState
                location: ~
            inputs:
              - SignalCall:
                  signal:
                    id: button_state
                    scope: Input
                  typing:
                    Enumeration: Button
                  location: ~
                  dependencies: ~
              - SignalCall:
                  signal:
                    id: counter
                    scope: Local
                  typing: Integer
                  location: ~
                  dependencies: ~
              - SignalCall:
                  signal:
                    id: reset_limit_1
                    scope: Input
                  typing: Integer
                  location: ~
                  dependencies: ~
            typing:
              Enumeration: ResetState
            location: ~
            dependencies: ~
        location: ~
    unitary_nodes: {}
    location: ~
    graph: ~
component: ~
location: ~

